lib/task/binder/tms/arFetchTmsV2.class.php

```diff
@@ -0,0 +1,430 @@
+            $creation['date'] = $value;
+
+            break;
+
+          // TODO: Allow multiple artist import
+          case 'artistList':
+            if (count($value) > 0)
+            {
+              if (isset($value[0]['displayName']))
+              {
+                $creation['actorName'] = $value[0]['displayName'];
+              }
+
+              if (isset($value[0]['displayDate']))
+              {
+                $creation['actorDate'] = $value[0]['displayDate'];
+              }
+            }
+
+            break;
+
+          // Digital object
+          case 'fullImage':
+            // Replace and combine backslashes with forwardslashes
+            $path = preg_replace('/\/+/', '/', str_replace('\\', '/', $value));
+
+            // Clear existing digital object
+            if (null !== $digitalObject = $artwork->getDigitalObject())
+            {
+              $digitalObject->delete();
+            }
+
+            // Create a new one
+            try
+            {
+              $filename = basename($path);
+              $contents = file_get_contents($path);
+
+              $digitalObject = new QubitDigitalObject;
+              $digitalObject->informationObjectId = $artwork->id;
+              $digitalObject->assets[] = new QubitAsset($filename, $contents);
+              $digitalObject->usageId = QubitTerm::MASTER_ID;
+              $digitalObject->indexOnSave = false;
+              $digitalObject->save();
+
+              // Add full path as property
+              $this->addOrUpdateProperty('FullImage', $digitalObject->getFullPath(), $artwork);
+
+              // If a thumbnail has been created, save as a property too and as a
+              // global variable to use it when the components are created/updated.
+              if (isset($digitalObject->thumbnail))
+              {
+                $this->artworkThumbnail = $digitalObject->thumbnail->getFullPath();
+                $this->addOrUpdateProperty('Thumbnail', $this->artworkThumbnail, $artwork);
+              }
+            }
+            catch (Exception $e)
+            {
+              $this->logger->info('arFetchTms - Error importing artwork image from "'.$value.'": '.$e->getMessage());
+            }
+
+            break;
+
+          case 'terms':
+            $content = array();
+            foreach ($value as $term)
+            {
+              if (!empty($term['path']))
+              {
+                // Remove redundant part of breadcrumb
+                $content[] = str_replace('Authorities > Attributes > Objects > ', '', $term['path']);
+              }
+            }
+
+            $this->addOrUpdateNote(QubitTerm::GENERAL_NOTE_ID, implode($content, "; "), $artwork);
+
+            break;
+        }
+      }
+    }
+
+    $artwork->save();
+
+    if (count($creation))
+    {
+      $this->addOrUpdateCreationEvent($creation, $artwork);
+    }
+  }
+
+  private function getTmsComponentsData($artwork, $componentsParent)
```

Comment:
I like the decomposition of the `getTmsComponentsData` functionality into it's own method.  Is it possible to do the same for the V1 API to make it easier to compare the two?

Reply:
I didn't try to normalize and abstract that part because the process to get the components data is quite different between versions:

In the v1, the TMS Component IDs are obtained from the TMS Object response and they are fetched one by one using those IDs. Whoever, in v2 the TMS Object response doesn't contain the TMS Components IDs and the only way to get them is in a single request which includes all the metadata for all of them.

Please, let me know if you still one to normalize a little more the process and I'll give it a try.

Reply:
Ah okay, I think it's fine to leave this as it is then. :)